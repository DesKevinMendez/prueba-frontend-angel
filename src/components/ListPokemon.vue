<template>

    <div :key="order" class="bg-blue-100 mx-auto w-2/3 flex flex-col mt-5">

        <input type="text" 
        class="bg-white border-2 border-blue-500 p-2" 
        placeholder="Escribir nombre del Pokemon"
        v-model="term"
        >

        <!-- <label class="relative inline-flex items-center mr-5 cursor-pointer mt-5 ml-5" >
            <input @click="changeOrder" type="checkbox" value="" class="sr-only peer" checked>
                <div class="w-11 h-6 bg-gray-200 rounded-full peer dark:bg-gray-700 peer-focus:ring-4 peer-focus:ring-green-300 dark:peer-focus:ring-green-800 peer-checked:after:translate-x-full peer-checked:after:border-white after:content-[''] after:absolute after:top-0.5 after:left-[2px] after:bg-white after:border-gray-300 after:border after:rounded-full after:h-5 after:w-5 after:transition-all dark:border-gray-600 peer-checked:bg-green-600"></div>
            <span class="ml-3 text-sm font-medium text-gray-900 dark:text-gray-300">Ordenar alfabeticamente</span>
        </label> -->

        <button @click="changeOrder" class="bg-blue-500 w-1/4 mt-5 mx-auto hover:bg-blue-700 text-white font-bold py-2 px-4 rounded-full">
            Ordenar
        </button>

        <div v-if="order" class="overflow-y-auto h-50  mt-5">
            <Item 
                v-for="poke in pokemonByTerm"
                :key="poke"
                :poke="poke"
            />
        </div>

        <div v-if="!order" class="overflow-y-auto h-50  mt-5">
            <Item 
                v-for="poke in orderByAlphabetic"
                :key="poke"
                :poke="poke"
            />
        </div>

    </div>

</template>

<script>
    import { defineAsyncComponent } from 'vue';
    import { mapGetters } from 'vuex';

export default{

    data(){
        return{
            term: '',
            order: false,
        }
    },

    components: {
        Item: defineAsyncComponent( () => import('./PokemonSelected.vue'))
    },
    computed:{
        ...mapGetters('pokemonStore',['getPokemonsByTerm','getPokemonsById', 'getPokemonsByOrder']),
        pokemonByTerm(){
            return this.getPokemonsByTerm(this.term)
        },

        orderByAlphabetic(){
            return this.getPokemonsByOrder(this.term)
        }
        

    }, 

    methods:{
        changeOrder(){
            this.order = !this.order
            console.log("Se hizo clic", this.order)
            return
        }
    },
    mounted(){
        this.changeOrder()
    },

    watch:{
        order(){
            
        }
    }

}

</script>