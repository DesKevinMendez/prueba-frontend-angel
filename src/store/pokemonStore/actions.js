import pokemonApi from "@/api/pokemonApi"


export const LoadPokemons = async( {commit} ) => {

    let pokemonArray = []

    for (let i = 1; i < 100; i++) {
        const {data} = await pokemonApi.get(`/${i}`)
        pokemonArray.push({
            data
        })
    } 

    const entries = []

    for(let id of Object.keys(pokemonArray)) {

        let datos = pokemonArray[id]
        let {data} = datos
        let namePokemon = data.name
        let idPokemon =  data.id+""

        entries.push({
            id: idPokemon, name: namePokemon

        })
    }

    commit('setPokemons', entries)
}
