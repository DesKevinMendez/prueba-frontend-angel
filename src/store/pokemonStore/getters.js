
export const  getPokemonsByTerm = (state) => (term = '') =>{

    if( term.length === 0) return state.pokemons

    return state.pokemons.filter(namePokemon => namePokemon.name.toLowerCase().includes(term.toLocaleLowerCase() ))
    
}


export const  getPokemonsByOrder = (state) => (term = '') =>{

    if( term.length === 0) {
        const pokemonsOrdenados = state.pokemons.sort((a,b) => a.name.localeCompare(b.name))
        return pokemonsOrdenados
    } 


    const pokemonsFiltrados = state.pokemons.filter(namePokemon => namePokemon.name.toLowerCase().includes(term.toLocaleLowerCase() ))
    const pokemonsOrdenados = pokemonsFiltrados.sort((a,b) => a.name.localeCompare(b.name))
    console.log("DESDE GETERS¿¿¿¿¿¿¿¿¿¿¿¿",pokemonsOrdenados)
    return pokemonsOrdenados

}

export const  getPokemonsById = (state) => ( id = '' ) =>{
    
    const poke = state.pokemons.find( poke => poke.id === id )

    if(!poke)  return

    return {...poke}

}


