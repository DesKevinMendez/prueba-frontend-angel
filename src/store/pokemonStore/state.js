
 export default () => ({

   isLoading: false,
    pokemons: [

     ]
 })