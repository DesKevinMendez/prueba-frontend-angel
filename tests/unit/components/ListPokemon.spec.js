import { shallowMount } from "@vue/test-utils"
import { createStore } from "vuex"
import ListPokemon from '@/components/ListPokemon'
import pokemonStore from "@/store/pokemonStore"
import { pokemonState } from "../mock-data/test-pokemon-state"


const createVuexStore = (initialState) => 
createStore({
    modules: {
        pokemonStore: {
            ...pokemonStore,
            state: {...initialState}
        }
    }
    
})

describe( "Component: ListPokemon", () => {

    const store = createVuexStore(pokemonState)
    const mockRouter = {
        push: jest.fn()
    }


    let wrapper

    beforeEach( () => {
        jest.clearAllMocks()
        wrapper = shallowMount(ListPokemon, {
            global: {
                mocks:{
                    $router: mockRouter
                },
                plugins: [store]
            }
        })
    })


    test( "Debe llamar a getPokemonsByTerm sin termino de busqueda", () => {

        expect( wrapper.findAll('item-stub').length).toBe(3)
        expect( wrapper.html()).toMatchSnapshot()

    })

    test( "Debe llamar a getPokemonsByTerm y filtrar con el termino de busqueda", async() => {

        const input = wrapper.find('input')
        await input.setValue('ivy')

        expect( wrapper.findAll('item-stub').length).toBe(1)

    })




})