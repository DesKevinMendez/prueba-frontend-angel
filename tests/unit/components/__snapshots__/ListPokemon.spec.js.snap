// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: ListPokemon Debe llamar a getPokemonsByTerm sin termino de busqueda 1`] = `
<div class="bg-blue-300 mx-auto w-2/3 flex flex-col mt-5"><input type="text" class="bg-blue-100 border-2 border-blue-600 p-2" placeholder="Escribir nombre del Pokemon"><!-- <button @click="!orderChange"> Ordenar Alfabeticamente </button> -->
  <div class="overflow-y-auto h-40 mt-5">
    <item-stub poke="[object Object]"></item-stub>
    <item-stub poke="[object Object]"></item-stub>
    <item-stub poke="[object Object]"></item-stub>
  </div>
</div>
`;

exports[`Component: ListPokemon Se debe llamar a getPokemonsByTerm sin termino 1`] = `
<div class="bg-blue-300 mx-auto w-2/3 flex flex-col mt-5"><input type="text" class="bg-blue-100 border-2 border-blue-600 p-2" placeholder="Escribir nombre del Pokemon"><!-- <button @click="!orderChange"> Ordenar Alfabeticamente </button> -->
  <div class="overflow-y-auto h-40 mt-5">
    <item-stub poke="[object Object]"></item-stub>
    <item-stub poke="[object Object]"></item-stub>
    <item-stub poke="[object Object]"></item-stub>
  </div>
</div>
`;
